New Block: `core/tabs`

sethrubenstein opened the pull request:
## What?
This is Pew Research Center's approach to Tabs in the block editor and would fulfill much of the needs requested in https://github.com/WordPress/gutenberg/issues/34079. 

This is a new PR picking up where I left off on https://github.com/WordPress/gutenberg/pull/69256

This builds upon work started by @creativecoder in https://github.com/WordPress/gutenberg/pull/63689/ and expands on making this concept even more extensible for developers. 

The major changes from this original concept are the following:
- An overhaul of tab editing has been implemented. A Slotfill (thanks @fabiankaegy for the idea) is now used to manage the tabs list and individual tab items. This enhances the user experience in two key ways:
  - 1. Selecting a tab now highlights the entire block, making it easier to start editing.
  - 2. Perhaps more importantly, we're not monitoring changes to innerblocks content to manage the tabs list, which caused significant performance issues when there were more than 10 tabs. Additionally, with the new slotfill selection method, we no longer render the innerblocks if the tab is not selected. This also improves the overall editing performance of the block.
- Some attribute changes were made, notably the removal of tabIndex. We now add the tab index to each tab during render time. This change allows for greater programmatic usage of tabs and simplifies the process of copying and pasting new tab panes in the editor.
- Structural changes to the `core/tab` block will allow it to degrade more gracefully by enabling the `core/tabs` block to control all Interactivity API adoption. If a `core/tab` block is displayed outside of `core/tabs`, it should simply show the tab content without any issues or modifications.
- The addition of various tab state colors as attributes and color controls enhances the user interface. Now, there are color controls and simple styles to support tab hover, tab active, tab background, text, and more.
- The addition of "deep linking" support allows for URL activation of a specific tab for content sharing. By passing a desired tab's anchor id into the address bar the tabs block will auto open to that tab on page load and scroll so it is in the viewport. 


## Why?
As the block editor and the concept of "full site editing" matures, I'm a strong believer that some basic ui concepts and paradigms should be available in the block library. Tabs is one such basic ui concept, dialog/modals are another. 

## How?

## Testing Instructions
1. Open a post
2. Add a tabs block
3. To start, two sample tabs are generated; click the plus icon in the tabs list to create a new tab
4. Add content
5. Preview and ensure tabs are selectable.

### Testing Instructions for Keyboard

On the frontend, simply tab and hit enter to navigate through tabs. ~In the editor, unfortunately due to the Slotfill nature of selecting and activating tabs this is currently not possible.~  (This is now possible by moving the logic into the tab link itself see https://github.com/WordPress/gutenberg/pull/69789#issuecomment-2802236850)

## Screenshots or screencast

https://github.com/user-attachments/assets/ff340893-7abf-462e-be39-114cd33488f5

sethrubenstein committed:
Initial commit

sethrubenstein committed:
Moving back to a manually constructed list in the parent block

sethrubenstein committed:
Moving more toward iAPI for easier use

sethrubenstein committed:
Moving more toward iAPI for easier use

sethrubenstein committed:
Some final cleanup before submitting pr

sethrubenstein committed:
Fixes php linting errors and forgot to hook on style functions for core/tab

sethrubenstein committed:
Adding fixtures/tests updating block library version number

sethrubenstein committed:
Why is PHPCS reporting this as misformatted??

sethrubenstein committed:
Whoops mis-replaced this file

sethrubenstein commented:
Alright alright alright, all tests (minus 2 that I'm pretty sure are unrelated to my PR) have passed 😁 and I am ready for people to start reviewing and offer feedback on the new `core/tabs` and `core/tab` blocks.

sethrubenstein committed:
iAPI bug fix, merge artifact - ensuring dynamic styles generation works in favor of static

sethrubenstein committed:
iAPI bug fix, merge artifact - ensuring dynamic styles generation works in favor of static

sethrubenstein committed:
Fixing failed tabs fixtures and removing old readme files from prc-block-library

sethrubenstein committed:
PHPCS linting

sethrubenstein committed:
I was really overthinking that, untill I was up late overthinking it.
And I thought, well I'm not actually doing any attributes in the block markup that I wouldn't be aware of so I could make this cleaner and more importantly more effecient.

Infinite-Null commented on packages/block-library/src/tab/index.js:
```diff
@@ -0,0 +1,20 @@
+ * Internal dependencies
+ */
+import edit from './edit';
+import icon from './icon';
```
The tabs block looks really good. Great work!

I think this icon should be coming from `@wordpress/icons`. 

1. First you need to add your icon:
`packages/icons/src/library/your-icon-name.js`

2. Add export for icon here:
`packages/icons/src/index.js`
```js
export { default as yourIconName } from './library/your-icon-file-name';
```

3. Then you can use it here like:
```
/**
 * WordPress dependencies
 */
import { yourIcon as icon } from '@wordpress/icons';
```

#### This is how it is done in other core blocks 
#### References:
https://github.com/WordPress/gutenberg/blob/6985f08c079fdf5bee969c420e483e582dc1b37e/packages/block-library/src/archives/index.js#L4

https://github.com/WordPress/gutenberg/blob/6985f08c079fdf5bee969c420e483e582dc1b37e/packages/block-library/src/comments-pagination-next/index.js#L5

https://github.com/WordPress/gutenberg/blob/6985f08c079fdf5bee969c420e483e582dc1b37e/packages/block-library/src/file/index.js#L5

#### PR Reference:
https://github.com/WordPress/gutenberg/pull/29428/files#diff-6fe62f58426c1aa17ab2b8d54c1563553446a5b02343fa183f97d5ea98e2d912

Infinite-Null commented on packages/block-library/src/tabs/index.js:
```diff
@@ -0,0 +1,22 @@
+ */
+
+import edit from './edit';
+import icon from './icon';
```
The same should be done here.

sethrubenstein committed:
Accessibility improvements and cleanup

sethrubenstein committed:
Regen tabs fixtures

Mamaduka commented:
Thanks for contributing, @sethrubenstein!

> An overhaul of tab editing has been implemented. A Slotfill (thanks @fabiankaegy for the idea) is now used to manage the tabs list and individual tab items. 

What was an initial version of this? My worry here is that SlotFills are public APIs by default, so anyone who knows their `name` (and can read source code) can add anything to them. This can break the tab block or prevent future iterations. Similar public APIs limit our ability to make significant changes to the UI.

Mamaduka commented:
> Are those slot fills when implemented that way accessible to anyone?

Technically, yes. If you know the `name` key, you can inject your components by using the `Fill` component. The core uses `Symbol` for some private/experimental SlotFills to avoid leaking the APIs.

Example: https://github.com/WordPress/gutenberg/blob/99c0aaa45838ffb3234250252ee8b60fe2ee85a2/packages/editor/src/components/more-menu/view-more-menu-group.js#L8

I've yet to test this; I was just curious about design decisions. I think it's an interesting challenge, and I'm looking forward to seeing the final result :)

P.S. This also reminded me of the Tabs API designs by UI libraries. Example: https://ariakit.org/components/tab.

sethrubenstein commented:
> Thanks for contributing, @sethrubenstein!
> 
> > An overhaul of tab editing has been implemented. A Slotfill (thanks @fabiankaegy for the idea) is now used to manage the tabs list and individual tab items.
> 
> What was an initial version of this? My worry here is that SlotFills are public APIs by default, so anyone who knows their `name` (and can read source code) can add anything to them. This can break the tab block or prevent future iterations. Similar public APIs limit our ability to make significant changes to the UI.

That is an interesting point, I hadn't thought of that. I'll try the Symbol method to see if I can get that to work well and get back to you all later in the week.

sethrubenstein committed:
Cool, that slotfill key scoping solution from @Mamaduka works

sethrubenstein committed:
Moving tabs icons into icon library proper as requested by @Infinite-Null

sethrubenstein commented:
That SlotFill key scoping solution worked without issue @Mamaduka, I'll keep that one in mind whenever I need to make a "private" SlotFill. 

Also, updated the icons so they're in @wordpress/icons proper @Infinite-Null

alexstine commented:
@sethrubenstein Not really involved these days in WordPress but every now and then, I glance over the PRs.

> In the editor, unfortunately due to the Slotfill nature of selecting and activating tabs this is currently not possible.

If this PR is starting off by admitting that keyboard only users will not have access in the editor, maybe this type of UI doesn't belong in core. Just my two sense without testing everything. Maybe you've thought of a way around this.

sethrubenstein commented:
@alexstine Yeah this is something I definitely would like some feedback and testing on. At first I thought my keyboard handlers on the parent block were the only way to go, and they were not firing because of losing focus when I tried to navigate between tabs so I pulled them. I just went back in and changed out the <a/> tab link elements for <button/> elements and moved my handlers for when the user select’s the tab there and this is surprisingly working pretty well. 

I’m debating changing the frontend elements to be buttons instead of a tags as well if it were not for the href signal? Anyone else have another suggestions on keyboard handlers that should be present that currently are not?

sethrubenstein committed:
Some updates for better keyboard handler support

alexstine commented:
Seems like the tabs on the front end really don't work with screen readers. Needs to be adjusted to follow spec.

https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Reference/Roles/tab_role#example

Back-end more or less works now but I don't have time to inspect the code at the moment to see if improvements can be made.

FYI @afercia @joedolson

sethrubenstein commented:
@alexstine Thanks, I'll get that changed sometime tomorrow.

sethrubenstein committed:
#34079 Fixing accesibility on frontend tab render

sethrubenstein commented:
I had the wrong aria-role and some unneeded structure, fixed.

alexstine commented:
@sethrubenstein This is testing much better now. Nice.

mikachan commented:
Thanks for picking this up! I'd be happy to review this when you're ready.

luisherranz commented:
> I'm moving away from that approach. Instead, I'll be relying on iAPI global state and iAPI templates.

Is it your intention to do something similar to [this](https://github.com/WordPress/gutenberg/pull/62906/commits/1077e73ccc93e06b13bbf8254a9fd8968d6c361b)? 

> `$tab_id = wp_unique_id( 'tab__' );`

We should look at what happens when there is a client-side navigation to make sure this block is compatible. Maybe we should use `getServerState()` to replace the ids, although depending on how you structure the state, it could be reset unintentionally, for example, by changing the tab that the user has open at that moment. So maybe we should try to make this id stable across renders, similar to the Query block one, which is saved as an attribute.

Anyway, just something to keep in mind as you move forward with that implementation.

sethrubenstein commented:
@luisherranz Good point on the unique id, I'll double check that on client side navigation. Otherwise I'll build a more reproducible id from the tab label slug. Expect another push on this PR with changes to how I'm currently handling the tabs list shortly after WCUS.

sethrubenstein committed:
Refactor tabs, pulling the latest out of prc-block-library/tabs for Gutenberg Core.

These changes bring in the following functionality:
- Tab labels can contain richtext now. We strip html tags when adding these to the iAPI.
- Issues with React focus issues in the editor have been resolved.
- Added ContrastChecker support to all color pairings.
- Updated default styles for tabs, links, button
- Better styles support for colors and block gap
- Focus on making tabs as extensible by 3rd party devs as possible, added a browser event to signal to other js applications that tabs are initialized (we use this to add react component driven content to our IAPI driven tabs)

sethrubenstein commented:
Okay updates are now merged in from prc-block-library. Included in this PR is a slight refactor of our Tabs block.

The latest commit brings in these updates:
- Tab labels can contain richtext and line breaks now. We strip html tags when adding these to the iAPI.
- Issues with React focus issues in the editor have been resolved.
- Added ContrastChecker support to all color pairings.
- Updated default styles for tabs, links, button. These are early passes, totally open to any design suggestions here.
- Better styles support for colors and block gap
- Focus on making tabs as extensible by 3rd party devs as possible, added a browser event to signal to other js applications that tabs are initialized (we use this to add react component driven content to our IAPI driven tabs).

One question, should I update these blocks to be set to `experimental`?

sethrubenstein committed:
Update included context to match core block library pattern

sethrubenstein commented:
I'll have the experimental updates in by end of week.

sethrubenstein committed:
Making tabs & tab blocks experimental

sethrubenstein committed:
Merge branch 'trunk' into core/tabs

sethrubenstein committed:
Fix tabs experimental placement

sethrubenstein committed:
Pull custom event. Per work on core/dialog

mikachan commented on lib/blocks.php:
```diff
@@ -121,6 +121,8 @@ function gutenberg_reregister_core_block_types() {
 				'site-tagline.php'                 => 'core/site-tagline',
 				'site-title.php'                   => 'core/site-title',
 				'tag-cloud.php'                    => 'core/tag-cloud',
+				'tabs.php'                         => 'core/tabs',
```
These blocks are listed alphabetically, so it would be good to include `tab` and `tabs` in the correct alphabetical order here.

mikachan commented on packages/block-library/src/tab/add-tab-toolbar-control.js:
```diff
@@ -0,0 +1,48 @@
+					label={ __( 'Add Tab' ) }
+					onClick={ addTab }
+					showTooltip
+					text="Add Tab"
```
```suggestion
					text={ __( 'Add Tab' ) }
```

mikachan commented on packages/block-library/src/tab/edit.js:
```diff
@@ -0,0 +1,248 @@
+
+	const { anchor, label } = attributes;
+
+	// Focus the label RichText component when no label exists
+	// and when the block is mounted.
+	useEffect( () => {
+		if ( ! label && labelRef.current ) {
+			const timeoutId = setTimeout( () => {
+				labelRef.current.focus();
+			}, 100 ); // A really quick millisecond delay to ensure the ref is and block is selected and active.
+			timeoutRef.current = timeoutId;
+			return () => clearTimeout( timeoutRef.current );
+		}
+	}, [ label ] );
+
+	// Clean up timeouts on unmount
+	useEffect( () => {
+		return () => {
+			if ( timeoutRef.current ) {
+				clearTimeout( timeoutRef.current );
+			}
+		};
+	}, [] );
```
I'm wondering if we can handle this in another way rather than using a timeout. Did you consider using `requestAnimationFrame` here instead? Something like:

```
		requestAnimationFrame( () => {
			if ( labelRef.current ) {
				labelRef.current.focus();
			}
		} );
```

(I haven't tried this, I'm not sure if it would work the same.)

mikachan commented on packages/block-library/src/tabs/controls.js:
```diff
@@ -0,0 +1,233 @@
+			<InspectorControls>
+				<PanelBody title={ __( 'Tabs Settings' ) }>
+					<ToggleControl
+						label="Vertical Tabs"
```
```suggestion
						label={ __( 'Vertical Tabs' ) }
```

mikachan commented on packages/icons/src/index.ts:
```diff
@@ -281,6 +281,8 @@ export { default as tableRowBefore } from './library/table-row-before';
 export { default as table } from './library/table';
 export { default as tag } from './library/tag';
+export { default as tab } from './library/tab';
+export { default as tabs } from './library/tabs';
```
I think these should be in alphabetical order as well.

mikachan commented on packages/icons/src/library/tab.js:
```diff
@@ -0,0 +1,19 @@
+			d="M5.5498 10.3501V6.3501H9.8498V10.3501H11.3498V6.1001C11.3498 5.40974 10.7902 4.8501 10.0998 4.8501H5.2998C4.60945 4.8501 4.0498 5.40974 4.0498 6.1001V10.3501H5.5498ZM20 12.6001H4V14.1001L20 14.1001V12.6001ZM14 17.1001H4V18.6001H14V17.1001Z"
+		/>
+	</SVG>
+);
```
We should keep the icons local to the block whilst the block is still experimental, similar to how we're handling the icons for the Accordion block: https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/accordion/icon.js. They can be moved into the icons library as soon as the block is stabilised.

mikachan reviewed (commented):
Thanks so much for working on this, @sethrubenstein!

I've given this an initial test and read-through, and left some inline comments. It's testing really well so far 🙌

sirreal commented on packages/block-library/src/tabs/index.php:
```diff
@@ -0,0 +1,241 @@
+ * @return string Updated HTML.
+ */
+function block_core_tabs_render_block_callback( array $attributes, string $content, \WP_Block $block ): string {
+	wp_enqueue_script_module( '@wordpress/block-library/tabs/view' );
+
```
Regarding the module, I think there's some confusing behavior.

Here, it's manually enqueued. There's also a `viewScriptModule` which should automatically enqueue the module:

https://github.com/pewresearch/gutenberg/blob/9684714c01b60f94c294354f733ad38c34fc52fb/packages/block-library/src/tabs/block.json#L149

However that `file:./…` form can be problematic to work across Gutenberg and Core.

I'd like to remove this enqueue and use `@wordpress/block-library/tabs/view` for the `viewScriptModule` similar to what was done in #71742.

I'll leave a corresponding suggestion for the `viewScriptModule`.


```suggestion
```

---

These Core blocks and modules have a confusing build/registration setup that most extenders don't have to deal with, largely from needing to work seamlessly in the context of the Gutenberg plugin and in Core.

sirreal commented on packages/block-library/src/tabs/block.json:
```diff
@@ -0,0 +1,150 @@
+	"editorScript": "file:./index.js",
+	"editorStyle": "file:./index.css",
+	"style": "file:./style-index.css",
+	"viewScriptModule": "file:./view.js"
```
```suggestion
	"viewScriptModule": "@wordpress/block-library/tabs/view"
```

sethrubenstein committed:
Resolving comments in review

sethrubenstein committed:
Moving to requestAnimationFrame and cancelAnimationFrame for cleaning up requests instead of timeout.

sethrubenstein committed:
alphabetical order

sethrubenstein committed:
alphabetical order

sethrubenstein committed:
Merge branch 'trunk' into core/tabs

sethrubenstein committed:
Refactor color styles function to remove unnecessary parameter

sethrubenstein committed:
Merge branch 'core/tabs' of github.com:pewresearch/gutenberg into core/tabs

sethrubenstein committed:
List order

sethrubenstein committed:
list order

sethrubenstein committed:
Order

sethrubenstein committed:
Order

sethrubenstein commented:
Pulled in latest from trunk and resolved variety of code styling/formatting comments.

mikachan commented on packages/block-library/src/tab/controls.js:
```diff
@@ -0,0 +1,67 @@
+						__nextHasNoMarginBottom
+					/>
+					<ToggleControl
+						label={ __( 'Is Default Tab' ) }
```
```suggestion
						label={ __( 'Default Tab' ) }
```

I think this wording works better for a toggle label.

mikachan commented on packages/block-library/src/tab/edit.js:
```diff
@@ -0,0 +1,255 @@
+	// Focus the label RichText component when no label exists
+	// and when the block is mounted.
+	useEffect( () => {
+		if ( ! label && labelRef.current ) {
+			const animationId = requestAnimationFrame( () => {
+				if ( labelRef.current ) {
+					labelRef.current.focus();
+				}
+			} );
+
+			focusRef.current = animationId;
+			return () => cancelAnimationFrame( focusRef.current );
+		}
+	}, [ label ] );
+
+	// Clean up animation frames on unmount
+	useEffect( () => {
+		return () => {
+			if ( focusRef.current ) {
+				cancelAnimationFrame( focusRef.current );
+			}
+		};
+	}, [] );
```
In my testing, the first tab label is not automatically focused when the Tabs block is inserted. I think this must be because the `label` always exists, i.e. it defaults to "Tab 1" when first inserted. I think we should ensure this label is focused on first insertion, even if the label is not empty. What do you think?

mikachan commented on packages/block-library/src/tab/index.js:
```diff
@@ -0,0 +1,24 @@
+/**
+ * WordPress dependencies
+ */
+
```
```suggestion
```

Looks like this isn't needed here.

mikachan commented on packages/block-library/src/tabs/index.php:
```diff
@@ -0,0 +1,239 @@
+/**
+ * Registers the `core/tabs` block on the server.
+ *
+ * @since 6.8.0
```
```suggestion
 * @since 6.9.0
```

mikachan commented on packages/block-library/src/tab/index.php:
```diff
@@ -0,0 +1,129 @@
+ *
+ * @hook init
+ *
+ * @since 6.8.0
```
```suggestion
 * @since 6.9.0
```

mikachan commented on packages/block-library/src/tabs/editor.scss:
```diff
@@ -0,0 +1,33 @@
+.wp-block-tabs.wp-block {
+	&__content > .block-list-appender {
+		top: -38px;
```
It would be good to add a comment here explaining what this value relates to.

mikachan commented on packages/block-library/src/tabs/index.js:
```diff
@@ -0,0 +1,25 @@
+/**
+ * WordPress dependencies
+ */
+
```
```suggestion
```

mikachan reviewed (commented):
Thanks for the updates, @sethrubenstein! I've left some more inline comments after another round of testing.

It looks like the test fixtures need regenerating, and there are a couple of coding standards checks that are failing.

Edit: Actually, I think we need to remove the test fixtures completely until the block is stabilised, as experimental blocks aren't included in the integration tests.

t-hamano commented:
From what I can see, further code review, design review, and accessibility review are needed.

I propose punt this block until after 7.0.

joedolson commented:
I assume you mean postponing until 7.0, not *after* 7.0, and assuming that's true, I concur, @t-hamano

t-hamano commented:
@joedolson I meant "in a future release". But of course, we should aim to release this block in 7.0.

sethrubenstein committed:
Updates:
- Merge in trunk
- Resolved unneeded test fixtures https://github.com/WordPress/gutenberg/pull/69789#pullrequestreview-3298386250
- Updated label https://github.com/WordPress/gutenberg/pull/69789#discussion_r2401478928
- Working on focus effect and iapi store refactor for private store method from dialog now

sethrubenstein committed:
Updates

sethrubenstein committed:
Updated to private store method that we implemented in core/dialog

sethrubenstein committed:
Fix initial mount of tabs and focusing on first tab richtext

sethrubenstein committed:
Cleanup

sethrubenstein committed:
Cleanup

sethrubenstein commented:
Huzzah. Okay @mikachan your requested updates are in. 
- I made some changes around initial block insertion behavior to account for focusing the first tab that I think is a little cleaner experience. 
- I also updated the iAPI store to match how I'm handling things for core/dialog 3rd party extensibility per @luisherranz . 

I think the last main focus area for me is cleaning up keyboard navigation in the editor context.

sethrubenstein committed:
Remove tests for now

sethrubenstein committed:
Cleanup

sethrubenstein commented:
Updated video of initial block experience:

https://github.com/user-attachments/assets/9b59ee51-9f80-4702-b1be-daeb6fdefbee

sethrubenstein committed:
Cleanup for coding standards

sethrubenstein committed:
Move away from slotfill, can restore if desired.

Mamaduka commented on packages/block-library/src/tab/tabs-list.js:
```diff
@@ -0,0 +1,104 @@
+									'core/strikethrough',
+								] }
+								placeholder={
+									__( 'Tab' ) + ` ${ currentBlockIndex + 1 }…`
```
It's recommended to construct similar translation strings via `sprintf`.

Something like.
```js
// translators: %s: tab number.
sprintf( __( 'Tab %d', currentBlockIndex + 1 ), 
```

Mamaduka commented on packages/block-library/src/tab/tabs-list.js:
```diff
@@ -0,0 +1,104 @@
+									'core/bold',
+									'core/italic',
+									'core/strikethrough',
+								] }
```
```suggestion
								allowedFormats={ [
									'core/bold',
									'core/italic',
									'core/strikethrough',
								] }
```

We could replace the hardcoded limit with `withoutInteractiveFormatting`. See the Details block for example.

sethrubenstein committed:
Replaced dynamic tab label with sprintf, replaced hard coded formats list for withoutInteractiveFormatting.

sethrubenstein committed:
Make markup render correctly when not focused in tabslist

sethrubenstein committed:
Add StaticLabel component to TabsList for improved tab label rendering

mikachan reviewed (approved):
Thanks for addressing all the recent feedback, @sethrubenstein 🙇 This block is working really well for me. I think we should bring it in, and we can make any further changes iteratively.

luisherranz commented on packages/block-library/src/tabs/view.js:
```diff
@@ -0,0 +1,218 @@
+		 * @type {number|null}
+		 */
+		get tabIndex() {
+			return createReadOnlyProxy( privateState.tabIndex );
```
You don't need a proxy if it's not an array or an object.


```suggestion
			return privateState.tabIndex;
```

luisherranz commented on packages/block-library/src/tabs/view.js:
```diff
@@ -0,0 +1,218 @@
+		 * @type {boolean}
+		 */
+		get isActiveTab() {
+			return createReadOnlyProxy( privateState.isActiveTab );
```
Same here.


```suggestion
			return privateState.isActiveTab;
```

luisherranz commented on packages/block-library/src/tabs/view.js:
```diff
@@ -0,0 +1,218 @@
+		 * @type {Array}
+		 */
+		get tabsList() {
+			return createReadOnlyProxy( privateState.tabsList );
```
For arrays, we should also prevent the use of functions that mutate the original array, like `push`.

luisherranz commented on packages/block-library/src/tabs/view.js:
```diff
@@ -0,0 +1,218 @@
+			 *
+			 * @type {boolean}
+			 */
+			get isActiveTab() {
```
Maybe this getter could also be exposed publicly?

luisherranz reviewed (commented):
Looks good!

A few small comments about the public store:
